File: Recursion/permutation_based/q51_n_queens.py
# n 皇后问题研究的是如何将 n 个皇后放置在 n×n 的棋盘上，并且使皇后彼此之间不能相互攻击。
#
#
#
#  上图为 8 皇后问题的一种解法。
#
#  给定一个整数 n，返回所有不同的 n 皇后问题的解决方案。
#
#  每一种解法包含一个明确的 n 皇后问题的棋子放置方案，该方案中 'Q' 和 '.' 分别代表了皇后和空位。
#
#  示例:
#
#  输入: 4
# 输出: [
#  [".Q..",  // 解法 1
#   "...Q",
#   "Q...",
#   "..Q."],
#
#  ["..Q.",  // 解法 2
#   "Q...",
#   "...Q",
#   ".Q.."]
# ]
# 解释: 4 皇后问题存在两个不同的解法。
#
#
#
#
#  提示：
#
#
#  皇后，是国际象棋中的棋子，意味着国王的妻子。皇后只做一件事，那就是“吃子”。当她遇见可以吃的棋子时，就迅速冲上去吃掉棋子。当然，她横、竖、斜都可走一到七步
# ，可进可退。（引用自 百度百科 - 皇后 ）
#
#  Related Topics 回溯算法
#  👍 477 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    # 写法1：紧凑
    def solveNQueens(self, n: int) -> List[List[str]]:
        def dfs(path_cols, xy_dif, xy_sum):
            row = len(path_cols)
            if row == n:
                res.append(list(path_cols))
                return

            for col in range(n):  # 遍历每一列
                if col not in path_cols and row - col not in xy_dif and row + col not in xy_sum:
                    dfs(path_cols + [col], xy_dif + [row - col], xy_sum + [row + col])

        res = []
        dfs([], [], [])
        return [["." * i + "Q" + "." * (n - i - 1) for i in subres] for subres in res]

    # 法1(写法2)：拆分开
    def solveNQueens1_2(self, n: int) -> List[List[str]]:
        if n <= 0: return []
        res = []
        self.dfs(n, [], [], [], res)
        return self.drawBoard(n, res)

    def dfs(self, n, path_cols, xy_dif, xy_sum, res):
        def isValid():
            return col not in path_cols and \
                   row - col not in xy_dif and row + col not in xy_sum

        row = len(path_cols)
        if row == n:
            res.append(list(path_cols))
            return

        for col in range(n):  # 遍历每一列
            if not isValid(): continue
            self.dfs(n, path_cols + [col], xy_dif + [row - col], xy_sum + [row + col], res)

    def drawBoard(self, n, res):
        return [['.' * i + 'Q' + '.' * (n - i - 1) for i in subRes] for subRes in res]

    # 法2：# ↓不建议global，否则solveNQueens中应在执行dfs2前将它们clear
    # result = []
    # cols, pie,na = set(), set(), set()
    def solveNQueens2(self, n: int) -> List[List[str]]:
        if n < 1: return []
        self.result = []
        self.cols, self.pie, self.na = set(), set(), set()
        self.dfs2(n, 0, [])
        return [['.' * i + 'Q' + '.' * (n - i - 1) for i in states] \
                for states in self.result]

    def dfs2(self, n, row, cur_state):
        if row >= n:
            self.result.append(cur_state)
            return  # cur_state ↑: 保存每行的Q所在的col_idx

        for col in range(n):
            if col in self.cols or \
                    row + col in self.pie or row - col in self.na:
                continue
            self.cols.add(col)
            self.pie.add(row + col)
            self.na.add(row - col)
            self.dfs2(n, row + 1, cur_state + [col])
            self.cols.remove(col)
            self.pie.remove(row + col)
            self.na.remove(row - col)

# leetcode submit region end(Prohibit modification and deletion)

File: Recursion/permutation_based/q51_n_queens.java
package Recursion.permutation_based;

import java.util.*;

public class q51_n_queens {
    private Set<Integer> cols, xyDiff, xySum; // 主对角xy_diff, 副对角xy_sum
    private int n;
    private List<List<String>> res;

    // 法2：❤回溯--传入Deque<Integer> paths（传入每行的Q_col）
    public List<List<String>> solveNQueens(int n) {
        // 设置成员变量，减少参数传递，具体作为方法参数还是作为成员变量，请参考团队开发规范
        this.n = n;
        cols = new HashSet<>();
        xyDiff = new HashSet<>();
        xySum = new HashSet<>();
        res = new ArrayList<>();

        // 二维棋盘转一维（只保存每行Q_col）
        Deque<Integer> paths = new ArrayDeque<>();
        dfs(n, 0, paths);// 深搜 -- 行row_i
        return res;
    }

    private void dfs(int n, int row, Deque<Integer> paths) {
        if (row == n) {
            res.add(convert(paths));
            return;
        }
        // 固定当前row_i, 遍历列col_j
        for (int col = 0; col < n; col++) {
            if (!check(row, col)) continue;
            cols.add(col); xyDiff.add(row-col); xySum.add(row+col);
            paths.addLast(col);
            dfs(n, row+1, paths);
            paths.removeLast();
            cols.remove(col); xyDiff.remove(row-col); xySum.remove(row+col);
        }
    }

    private boolean check(int row, int col) {
        return !(cols.contains(col) || xyDiff.contains(row-col)
                || xySum.contains(row+col));
    }

    private List<String> convert(Deque<Integer> paths) {
        List<String> chessboard = new ArrayList<>();
        for (Integer colQ: paths) {
            // ❤ String -- 【".".repeat(n)】 等价py: "."*n ↓
            StringBuilder row = new StringBuilder(".".repeat(n));
            row.replace(colQ, colQ+1, "Q");
            chessboard.add(new String(row));// 共后缀n行(棋盘)
        }					// ↑StringBuilder转String
        return chessboard;
    }

    // 法1【不荐，String复杂】：传入String[] chessboard
    public List<List<String>> solveNQueens2(int n) {
        this.n = n;
        cols = new HashSet<>();
        xyDiff = new HashSet<>();
        xySum = new HashSet<>();
        res = new ArrayList<>();

        // 初始化棋盘(看做二维path)
        // 法1【荐】:
        String[] chessboard = new String[n]; // 最后
        String row = ".".repeat(n); // py: "."*n
        for (int i = 0; i < n; i++) {
            Arrays.fill(chessboard, row);// 按序填充
        }

        // 法2[replace需要StringBuilder，转换太麻烦]:
        // StringBuilder row = new StringBuilder(".".repeat(n)); // py: "."*n
        // List<StringBuilder> chessboard = Collections.nCopies(n, row); //❤❤等价py: [row]*n

        dfs(n, 0, chessboard);
        return res;
    }

    private void dfs(int n, int row, String[] chessboard) {
        if (row == n) {
            List<String> newborad = Arrays.asList(chessboard);
            res.add(new ArrayList<>(newborad)); // 不可直接传入asList(引用), 最终结果全为'....'
            return;
        }
        for (int col = 0; col < n; col++) {
            if (isValid(row, col)) {
                StringBuilder newRow = new StringBuilder();
                newRow.append(".".repeat(n)); // Math.max(0, n),共后缀n行(棋盘)
                newRow.replace(col, col+1, "Q");
                String originRow = chessboard[row];

                chessboard[row] = new String(newRow);
                cols.add(col); xyDiff.add(row-col); xySum.add(row+col);
                dfs(n, row+1, chessboard);
                cols.remove(col); xyDiff.remove(row-col); xySum.remove(row+col);
                chessboard[row] = originRow;
            }
        }
    }

    private boolean isValid(int row, int col) {
        if (cols.contains(col) || xyDiff.contains(row-col)
                || xySum.contains(row+col)) {
            return false;
        }
        return true;
    }




}
